shared/metadata.go

```diff
@@ -20,3 +50,94 @@ type MetadataLink struct {
 	TestPath string `yaml:"test"`
 	URL      string
 }
+
+func (m MetadataResults) Len() int           { return len(m) }
+func (m MetadataResults) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
+func (m MetadataResults) Less(i, j int) bool { return m[i].String() < m[j].String() }
+
+func (m MetadataResult) String() string {
+	var urlString string
+	for _, url := range m.URLs {
+		urlString += url
+	}
+	return m.Test + urlString
+}
+
+// GetMetadataResponse retrieves the response to a WPT Metadata query.
+func GetMetadataResponse(testRuns []TestRun, client *http.Client) MetadataResponse {
+	metadataByteMap, err := util.CollectMetadata(client)
+	if err != nil {
+		return MetadataResponse{}
+	}
+	metadata := parseMetadata(metadataByteMap)
+	return constructMetadataResponse(testRuns, metadata)
+}
+
+// parseMetadata collects and parses all META.yml files from
+// wpt-metadata reposiroty.
+func parseMetadata(metadataByteMap map[string][]byte) map[string]Metadata {
+	var metadataMap = make(map[string]Metadata)
+	for path, data := range metadataByteMap {
+		var metadata Metadata
+		err := yaml.Unmarshal(data, &metadata)
+		if err != nil {
+			log.Warningf("Failed to unmarshal %s.yml.", path)
+			continue
+		}
+		metadataMap[path] = metadata
+	}
+	return metadataMap
+}
+
+// ConstructMetadataResponse constructs the response to a WPT Metadata query.
+// When parsing 'link' nodes, assume there is no mising information nor duplicates;
+// assume each test for each browser type is only associated with one bug.
+func constructMetadataResponse(testRuns []TestRun, metadata map[string]Metadata) MetadataResponse {
+	res := MetadataResults{}
+	for folderPath, data := range metadata {
+		testMap := make(map[string][]string)
+
+		for _, link := range data.Links {
+			var urls []string
+
+			var fullTestName = ""
+			if strings.HasPrefix(link.TestPath, folderPath) {
+				fullTestName = link.TestPath
+			} else {
+				fullTestName = path.Join(folderPath, link.TestPath)
+			}
+
+			if _, ok := testMap[fullTestName]; !ok {
+				testMap[fullTestName] = make([]string, len(testRuns))
```

Comment:
This is both incorrect and unnecessary.

1. The second param of `make` is the length, not the capacity. Yet we don't really know the length at this point, so we might end up with extra empty strings at the end of the slice.
2. We should make use of the zero value of slice and `append`. If a key does not exist in the map, the map will give you a zero value, which is a nil slice in this case. You can actually `append` to a nil slice without running into nil pointer errors.

Reply:
The reasoning here is because I want to return Metadata in a similar fashion to rest results, e.g. when querying for 4 browsers, the metadata response will look like 

test: "IndexedDB/bindings-inject-key.html"
urls: ["bugs.chromium.org/p/chromium/issues/detail?id=934844", "", "", ""]

while a test result looks like

test: "/2dcontext/building-paths/canvas_complexshapes_arcto_001.htm"
legacy_status: [{passes: 0, total: 1}, {passes: 0, total: 1}, {passes: 0, total: 1}, {passes: 0, total: 1}

So if there are empty strings, it just means the corresponding browser for this test does not have any bug associated with it. Len(testRuns) gives me both the length and the capacity in this case.

Reply:
Ah I see. You're right!